policy.go

```diff
@@ -305,3 +454,230 @@ func readAndValidateLockNVIndexPublic(tpm *tpm2.TPMContext, index tpm2.ResourceC
+	}
+
+	if !staticInput.AuthorizeKeyPublic.NameAlg.Supported() {
+		return errors.New("public area of dynamic authorization policy signature verification key has an unsupported name algorithm")
+	}
+	authorizeKey, err := tpm.LoadExternal(nil, staticInput.AuthorizeKeyPublic, tpm2.HandleOwner)
+	if err != nil {
+		return xerrors.Errorf("cannot load public area for dynamic authorization policy signature verification key: %w", err)
+	}
+	defer tpm.FlushContext(authorizeKey)
+
+	h := staticInput.AuthorizeKeyPublic.NameAlg.NewHash()
+	h.Write(dynamicInput.AuthorizedPolicy)
+
+	authorizeTicket, err := tpm.VerifySignature(authorizeKey, h.Sum(nil), dynamicInput.AuthorizedPolicySignature)
+	if err != nil {
+		return xerrors.Errorf("dynamic authorization policy signature verification failed: %w", err)
+	}
+
+	if err := tpm.PolicyAuthorize(policySession, dynamicInput.AuthorizedPolicy, nil, authorizeKey.Name(), authorizeTicket); err != nil {
+		return xerrors.Errorf("dynamic authorization policy check failed: %w", err)
+	}
+
+	pinIndex.SetAuthValue([]byte(pin))
+	if _, _, err := tpm.PolicySecret(pinIndex, policySession, nil, nil, 0, hmacSession); err != nil {
+		return xerrors.Errorf("cannot execute PolicySecret assertion: %w", err)
+	}
+
+	lockIndex, err := tpm.CreateResourceContextFromTPM(lockNVHandle)
+	if err != nil {
+		return xerrors.Errorf("cannot obtain context for lock NV index: %w", err)
+	}
+	if err := tpm.PolicyNV(lockIndex, lockIndex, policySession, nil, 0, tpm2.OpEq, hmacSession); err != nil {
+		return xerrors.Errorf("policy lock check failed: %w", err)
+	}
+
+	return nil
+}
+
+// lockAccessToSealedKeysUntilTPMReset locks access to sealed key objects created by this package until the next TPM restart
+// or TPM reset (aka, Startup(CLEAR)). This works by enabling the read lock bit for a well-known NV index. The static authorization
+// policy of sealed key objects created by this package contain a TPM2_PolicyNV assertion on this index, which will fail once the
+// read lock bit has been enabled.
+func lockAccessToSealedKeysUntilTPMReset(tpm *tpm2.TPMContext, session tpm2.SessionContext) error {
+	handles, err := tpm.GetCapabilityHandles(lockNVHandle, 1, session.IncludeAttrs(tpm2.AttrAudit))
+	if err != nil {
+		return xerrors.Errorf("cannot obtain handles from TPM: %w", err)
+	}
+	if len(handles) == 0 || handles[0] != lockNVHandle {
+		// Not provisioned, so no keys to protect.
+		return nil
+	}
+	lock, err := tpm.CreateResourceContextFromTPM(lockNVHandle)
+	if err != nil {
+		return xerrors.Errorf("cannot obtain context for lock NV index: %w", err)
+	}
+	if err := tpm.NVReadLock(lock, lock, session); err != nil {
+		var hErr *tpm2.TPMHandleError
+		if xerrors.As(err, &hErr) && hErr.Index == 2 && hErr.Code() == tpm2.ErrorAttributes {
+			// Not provisioned with a valid lock NV index, so no keys created by this package can
+			// be unsealed on this TPM anyway.
+			return nil
+		}
+		return xerrors.Errorf("cannot readlock pin NV index: %w", err)
```

Comment:
s/readlock/lock for read/ ?

Reply:
I've updated this to "cannot lock pin NV index for reading".